kairos-prover/kairos-circuit-logic/src/transactions.rs

```diff
@@ -0,0 +1,541 @@
+        }
+    }
+
+    impl Default for AccountsState {
+        fn default() -> Self {
+            unreachable!("AccountsState should always be created with AccountsState::new()");
+        }
+    }
+
+    impl AccountsState {
+        pub fn build_trie(&self) -> (TrieRoot<NodeHash>, Rc<MemoryDb<Account>>) {
+            let mut account_trie =
+                AccountTrie::new_try_from_db(Rc::new(MemoryDb::empty()), TrieRoot::Empty).unwrap();
+
+            let mut hasher = Sha256::new();
+            for (public_key, account) in self.l2.accounts.iter() {
+                hasher.update(public_key.as_slice());
+                let key = &KeyHash::from_bytes(&hasher.finalize_fixed_reset().into());
+
+                account_trie.txn.insert(key, account.clone()).unwrap();
+            }
+
+            let root = account_trie.txn.commit(&mut DigestHasher(hasher)).unwrap();
+
+            let db = account_trie.txn.data_store.db().clone();
+
+            (root, db)
+        }
+
+        pub fn random_deposit(
+            &mut self,
+            sender: sample::Index,
+            recipient: sample::Index,
+            amount_sampler: sample::Index,
+        ) -> (L1Deposit, TxnExpectedResult) {
+            let sender = self.l1.sample_keys(sender);
+            let recipient = self.l2.sample_keys(recipient).deref().clone();
+
+            let l1_balance = self
+                .l1
+                .accounts
+                .get_mut(&sender)
+                .expect("sender does not have an l1 account in AccountsState");
+
+            let amount = if *l1_balance == 0 {
+                return (
+                    L1Deposit {
+                        recipient,
+                        amount: 0,
+                    },
+                    TxnExpectedResult::Failure,
+                );
+            } else {
+                amount_sampler.index(*l1_balance as usize) as u64 + 1
+            };
+
+            if *sender == recipient {
+                return (L1Deposit { recipient, amount }, TxnExpectedResult::Failure);
+            }
+
+            let l2_account = self
+                .l2
+                .accounts
+                .get_mut(&recipient)
+                .expect("recipient does not have an l2 account in AccountsState");
+
+            // if the deposit will fail don't change the test model state
+            match (
+                l1_balance.checked_sub(amount),
+                l2_account.balance.checked_add(amount),
+            ) {
+                (Some(l1_bal), Some(l2_bal)) => {
+                    *l1_balance = l1_bal;
+                    l2_account.balance = l2_bal;
+
+                    (L1Deposit { recipient, amount }, TxnExpectedResult::Success)
+                }
+                _ => {
+                    unreachable!("For now I am not testing the case where the deposit fails");
+                    // (L1Deposit { recipient, amount }, TxnExpectedResult::Failure)
+                }
```

Comment:
Testing for failures would expand the scope of the PR even more. I wanted to get the circuit logic merged to unblock other work before more rigorous testing, but ended up implementing most of the proptest random test model despite that. 

I had not intended this pr to have proptests at all. The proptest work was originally going to be follow-up because it's more complex than the actual circuit code. As it stands right now the test model matches only successful operations, failures would be different in a number of few known case.

Reply:
Removed whole impl, since I no longer use `AccountsState` as `Arbitrary::Parameters`.